### PR TITLE
Draft Heterogeneous Data Loading

### PR DESCRIPTION
**Pull request by mkolodner-sc**
**Scope of work done**

Where is the documentation for this feature?: N/A

Did you add automated tests or write a test plan?

***Updated Changelog.md?*** NO

***Ready for code review?:*** NO

### CODE REVIEW

File: python/gigl/distributed/dataset_factory.py [hunk 1]
```diff
@@ -104,8 +96,39 @@ def _load_and_build_partitioned_dataset(
...
+
+    # TODO (mkolodner-sc): Move this code block to transductive splitter once that is ready
+    if _ssl_positive_label_percentage is not None:
+        assert (
+            loaded_graph_tensors.positive_label is None
+            and loaded_graph_tensors.negative_label is None
+        ), "Cannot have loaded positive and negative labels when attempting to select self-supervised positive edges from edge index."
+        positive_label_edges: Union[torch.Tensor, Dict[EdgeType, torch.Tensor]]
+        if isinstance(loaded_graph_tensors.edge_index, abc.Mapping):
+            assert isinstance(splitter, HashedNodeAnchorLinkSplitter)
```
Comment (kmontemayor2-sc) [inline python/gigl/distributed/dataset_factory.py:108]: why do we need this?  is `assert is not None` also ok?

File: python/gigl/distributed/dataset_factory.py [hunk 2]
```diff
@@ -104,8 +96,39 @@ def _load_and_build_partitioned_dataset(
...
+            positive_label_edges = {}
+            for supervision_edge_type in splitter.supervision_edge_types:
```
Comment (kmontemayor2-sc) [inline python/gigl/distributed/dataset_factory.py:110]: btw I think it's a bit weird to enforce that the splitter knows about the edge types, more of a nit though.

File: python/gigl/distributed/dist_ablp_neighborloader.py [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+        if isinstance(input_nodes, tuple):
+            if supervision_edge_type is None:
+                raise ValueError(
+                    "When using heterogeneous ABLP, you must provide supervision_edge_types."
+                )
+            self._is_input_heterogeneous = True
+            node_type, node_ids = input_nodes
+            assert (
+                supervision_edge_type[0] == node_type
+            ), f"Label EdgeType are currently expected to be provided as `anchor_node_type`_`relation`_`supervision_node_type`, \
```
Comment (kmontemayor2-sc) [inline python/gigl/distributed/dist_ablp_neighborloader.py:148]: nit. if people provide pyg edge type this it won't be represented as the `src_type_relation_type_dst_type` form.

File: python/gigl/distributed/dist_ablp_neighborloader.py [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+                got supervision edge type {supervision_edge_type} with anchor node type {node_type}"
+            supervision_node_type = supervision_edge_type[2]
```
Comment (kmontemayor2-sc) [inline python/gigl/distributed/dist_ablp_neighborloader.py:150]: Does this require a specific edge dir? IIRC it came up that we may (and generally do) have a different edge dir for labels than the rest of the graph.

File: python/gigl/distributed/dist_ablp_neighborloader.py [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+            drop_last=drop_last,
+            with_edge=True,
+            collect_features=True,
+            with_neg=False,
+            with_weight=False,
+            edge_dir=dataset.edge_dir,
+            seed=None,  # it's actually optional - None means random.
+        )
+
+        self.data = dataset
```
Comment (kmontemayor2-sc) [inline python/gigl/distributed/dist_ablp_neighborloader.py:313]: nit. Add a note we copied this over from the GLT base?

Comment (kmontemayor2-sc) [reply python/gigl/distributed/dist_ablp_neighborloader.py:313]: And also what we changed?

Ditto generally for this class.

File: python/gigl/distributed/dist_ablp_neighborloader.py [hunk 6]
```diff
@@ -0,0 +1,447 @@
...
+    ) -> Data:
+        data.y_positive = positive_labels
+        if negative_labels is not None:
+            data.y_negative = negative_labels
+        if isinstance(data, HeteroData):
+            del data.num_sampled_edges[self._positive_label_edge_type]
+            del data._edge_store_dict[self._positive_label_edge_type]
+            if negative_labels is not None:
+                del data.num_sampled_edges[self._negative_label_edge_type]
+                del data._edge_store_dict[self._negative_label_edge_type]
```
Comment (kmontemayor2-sc) [inline python/gigl/distributed/dist_ablp_neighborloader.py:438]: nit. add comment?

File: python/gigl/distributed/dist_neighbor_sampler.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    SamplerOutput,
+)
+from graphlearn_torch.typing import EdgeType, NodeType
+from graphlearn_torch.utils import count_dict, merge_dict, reverse_edge_type
+
+from gigl.types.sampler import LabeledNodeSamplerInput
+from gigl.utils.data_splitters import PADDING_NODE
+
+
+class DistLinkPredictionNeighborSampler(DistNeighborSampler):
```
Comment (kmontemayor2-sc) [inline python/gigl/distributed/dist_neighbor_sampler.py:19]: let's add TODOS about how we can upstream this back up to GLT?

File: python/gigl/types/graph.py [hunk 8]
```diff
@@ -107,13 +108,17 @@ class LoadedGraphTensors:
     # Unpartitioned Negative Edge Label
     negative_label: Optional[Union[torch.Tensor, dict[EdgeType, torch.Tensor]]]
 
-    def treat_labels_as_edges(self) -> None:
+    def treat_labels_as_edges(self, edge_dir: Literal["in", "out"]) -> None:
```
Comment (kmontemayor2-sc) [inline python/gigl/types/graph.py:111]: btw if we run into needing to do a lot of asserts to get mypy to be happy  about this let's do `Union[str, Literal["in", "out"]]`?

File: python/gigl/types/graph.py [hunk 9]
```diff
@@ -403,3 +459,14 @@ def to_homogeneous(
         n = next(iter(x.values()))
         return n
     return x
+
+
+def reverse_edge_type(edge_type: EdgeType) -> EdgeType:
```
Comment (kmontemayor2-sc) [inline python/gigl/types/graph.py:464]: nit. `_EdgeType` and return PyG or GiGL edge type as needed?

File: python/gigl/types/sampler.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+
+class LabeledNodeSamplerInput(NodeSamplerInput):
+    def __init__(
+        self,
+        node: torch.Tensor,
+        input_type: Optional[Union[str, NodeType]],
+        positive_labels: torch.Tensor,
+        negative_labels: Optional[torch.Tensor],
+        supervision_node_type: Optional[Union[str, NodeType]],
```
Comment (kmontemayor2-sc) [inline python/gigl/types/sampler.py:16]: should we add a TODO about different positive/negative node types?

File: python/gigl/utils/data_splitters.py [hunk 11]
```diff
@@ -197,7 +198,20 @@ def __call__(
...
+            if self._should_convert_labels_to_edges and is_label_edge_type(
+                edge_type_to_split
+            ):
+                is_label = True
+            elif (
+                not self._should_convert_labels_to_edges
+                and edge_type_to_split in self._supervision_edge_types
+            ):
+                is_label = True
+            if not is_label:
```
Comment (kmontemayor2-sc) [inline python/gigl/utils/data_splitters.py:212]: btw why do we skip here? can we add a comment about it?

File: python/gigl/utils/data_splitters.py [hunk 12]
```diff
@@ -270,6 +284,10 @@ def __call__(
             val = nodes_to_select[num_train : num_val + num_train]  # 1 x num_val_nodes
             test = nodes_to_select[num_train + num_val :]  # 1 x num_test_nodes
             splits[anchor_node_type] = (train, val, test)
+        assert (
+            len(splits) > 0
+        ), f"Found no edge types to split from the provided edge index: {edge_index.keys()} using supervision edge types {self._supervision_edge_types}"
```
Comment (kmontemayor2-sc) [inline python/gigl/utils/data_splitters.py:289]: nit. `raise ValueError` instead?

File: python/gigl/distributed/dist_neighbor_sampler.py [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+        negative_labels = (
+            inputs.negative_labels.to(self.device)
+            if inputs.negative_labels is not None
+            else None
+        )
+
+        labeled_seeds = positive_labels[positive_labels != PADDING_NODE]
+        if negative_labels is not None:
+            negative_seeds = negative_labels[negative_labels != PADDING_NODE]
+            labeled_seeds = torch.cat((labeled_seeds, negative_seeds), dim=0)
```
Comment (kmontemayor2-sc) [inline python/gigl/distributed/dist_neighbor_sampler.py:38]: Do you think it's helpful to `del positive_labels, negative_labels; gc.collect()` after this?

File: python/gigl/distributed/dist_neighbor_sampler.py [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+        self.max_input_size: int = max(self.max_input_size, input_seeds.numel())
+        inducer = self._acquire_inducer()
+        is_hetero = self.dist_graph.data_cls == "hetero"
+        metadata = {"positive_labels": positive_labels}
```
Comment (kmontemayor2-sc) [inline python/gigl/distributed/dist_neighbor_sampler.py:42]: nit. type?

File: python/gigl/distributed/dist_neighbor_sampler.py (hunk 7 above)
Comment (kmontemayor2-sc) [inline python/gigl/distributed/dist_neighbor_sampler.py:19]: `DistLinkPredictionABLPNeighborSampler` or similar since this only works for ABLP?

File: python/gigl/types/graph.py (hunk 8 above)
Comment (mkolodner-sc) [reply python/gigl/types/graph.py:111]: Sounds good, I think for now we don't need to add too any asserts where this is used or in the function, but I'll keep in mind if this comes up in the future

File: python/gigl/distributed/dist_neighbor_sampler.py (hunk 13 above)
Comment (mkolodner-sc) [reply python/gigl/distributed/dist_neighbor_sampler.py:40]: The labels are still sticking around in the `metadata`, so there won't be any memory saved by using this here.
